File: lib/fd.pl
%%%% operations on file-descriptors and raw I/O


:- verbatim('#include "fd.c"').


write_bytes(_, []) :- !.
write_bytes(_, '') :- !.
write_bytes(FD, DATA) :-
	atom(DATA),
	!,
	foreign_call(raw_write(FD, DATA, N)),
	sub_atom(DATA, N, LEN, 0, REST),
	write_bytes(FD, REST).
write_bytes(FD, DATA) :-
	foreign_call(raw_write(FD, DATA, N)),
	skip_bytes(N, DATA, REST),
	!,
	write_bytes(FD, REST).


%% will return [] on EOF, and block after that

read_bytes(FD, LEN, BYTES) :-
	foreign_call(raw_read(FD, LEN, BYTES)).


%% probably generally useful, find out how this is usually called

skip_bytes(0, _, []).
skip_bytes(N, [_|R], R2) :-
	N2 is N - 1,
	!,
	skip_bytes(N2, R, R2).


%% open file-descriptor

open_fd(FD, MODE, STREAM) :-
	open_fd(FD, MODE, STREAM, []).

open_fd(FD, write, STREAM, OPTIONS) :-
	open_fd(FD, 0, "w", OPTIONS, [output, mode(write)], STREAM).
open_fd(FD, read, STREAM, OPTIONS) :-
	open_fd(FD, 1, "r", OPTIONS, [input, mode(read)], STREAM).
open_fd(FD, append, STREAM, OPTIONS) :-
	open_fd(FD, 0, "a", OPTIONS, [output, mode(append)], STREAM).

open_fd(FD, INPUT, MODE, [], DATA, STREAM) :-
	name(M, MODE),
	foreign_call(open_fd(FD, INPUT, M, [file_no(FD)|DATA], STREAM)).
open_fd(FD, INPUT, MODE, [type(text)|MORE], DATA, STREAM) :-
	open_fd(FD, INPUT, MODE, MORE, [type(text)|DATA], STREAM).
open_fd(FD, INPUT, MODE, [type(binary)|MORE], DATA, STREAM) :-
	append(MODE, "b", MODE2),
	open_fd(FD, INPUT, MODE2, MORE, [type(binary)|DATA], STREAM).
open_fd(FD, INPUT, MODE, [_|MORE], DATA, STREAM) :-
	open_fd(FD, INPUT, MODE, MORE, DATA, STREAM).


%% wait for I/O on file-descriptors

poll_fds(FDS, READY) :-
	poll_fds(FDS, -1, READY).

poll_fds(FDS, TIMEOUT, READY) :-
	foreign_call(poll_fds(FDS, TIMEOUT, READY)).


%% generally useful, but it seldom works for me...

set_stream_buffer(STR, BUF) :-
	foreign_call(set_stream_buffer(STR, BUF)).

File: lib/fd.c
/* some operations on file-descriptors (raw I/O) */


#include <poll.h>


PRIMITIVE(open_fd, X fd, X input, X mode, X data, X result) 
{
  int len;
  FILE *fp = fdopen(fixnum_to_word(fd), to_string(mode, &len));
  X port = PORT(fp, input, ONE, data);
  return unify(port, result);
}


PRIMITIVE(raw_read, X fd, X count, X bytes)
{
  int fno = fixnum_to_word(check_fixnum(fd));
  XWORD n = fixnum_to_word(check_fixnum(count));
  ensure_string_buffer(n, NULL);
  size_t total = read(fno, string_buffer, n);

  if(total == -1)
    system_error(strerror(errno));

  return unify(bytes, string_to_list(string_buffer, total));
}


PRIMITIVE(raw_write, X fd, X bytes, X written)
{
  int fno = fixnum_to_word(check_fixnum(fd));
  int len;
  XCHAR *ptr = to_string(bytes, &len);
  size_t total = write(fno, ptr, len);

  if(total == -1)
    system_error(strerror(errno));

  return unify(written, word_to_fixnum(total));
}


PRIMITIVE(poll_fds, X fdlist, X timeout, X rdylist)
{
  static struct pollfd fds[ 256 ];	/* max */
  int n = 0;

  while(fdlist != END_OF_LIST_VAL) {
    fds[ n ].fd = fixnum_to_word(check_fixnum(deref(slot_ref(fdlist, 0))));
    fds[ n ].events = POLLIN | POLLOUT;
    fdlist = deref(slot_ref(fdlist, 1));
    ++n;
  }

  int rn = poll(fds, n, fixnum_to_word(check_fixnum(timeout)));

  if(rn == -1) {
    if(errno == EINTR)
      return unify(rdylist, END_OF_LIST_VAL);

    system_error(strerror(errno));
  }

  //XXX does not check for full heap
  X lst = END_OF_LIST_VAL;

  for(int i = 0; n > 0; ++i) {
    if((fds[ i ].revents & (POLLERR | POLLIN | POLLOUT)) != 0)
      lst = PAIR(word_to_fixnum(fds[ i ].fd), lst);

    --n;
  }

  return unify(rdylist, lst);
}


PRIMITIVE(set_stream_buffer, X port, X buf)
{
  X stream = get_stream(port);
  XCHAR *buffer;
  int mode;
  size_t size;

  if(buf == ZERO) {
    buffer = NULL;
    mode = _IONBF;
    size = 0;
  } 
  else {
    buffer = malloc(size = fixnum_to_word(check_fixnum(buf)));
    mode = _IOFBF;
  }

  if(setvbuf(port_file(stream), buffer, mode, size) != 0)
    system_error(strerror(errno));

  return 1;
}
